clustering/web/undertow/src/main/java/org/wildfly/clustering/web/undertow/session/DistributableSession.java

```diff
@@ -89,6 +89,9 @@ public void requestDone(HttpServerExchange exchange) {
                     // Don't propagate exceptions at the stage, since response was already committed
                     UndertowClusteringLogger.ROOT_LOGGER.warn(e.getLocalizedMessage(), e);
                 }
+            } else if (this.batch.getState() != Batch.State.CLOSED) {
```

Comment:
When is this ever the case?

Reply:
This was the initial problem in this issue. The first request invalidated the session and the second one detected it as invalid and threw an exception, see [this lines](https://github.com/wildfly/wildfly/blob/master/clustering/web/undertow/src/main/java/org/wildfly/clustering/web/undertow/session/DistributableSession.java#L254). The lock here was not released and the session remained locked. Remember I first added the release in the same `validate` method (before the throw) and you recommended me to move it to the `requestDone` method.

Reply:
Oh right.  Now I remember. :)
However, you'll want to make sure to resume the batch before closing.
e.g.
```java
try (BatchContext context = batcher.resumeBatch(this.batch)) {
    // Ensure batch is closed.
    this.batch.close();
}
```

Reply:
Right! Like in the previous if. It's done now. Just submitted the new version.